src/gz/write.rs

```diff
@@ -180,3 +182,273 @@ impl<W: Write> Drop for GzEncoder<W> {
+    ///
+    /// Attempts to write data to this stream may result in a panic after this
+    /// function is called.
+    ///
+    /// # Errors
+    ///
+    /// This function will perform I/O to finish the stream, returning any
+    /// errors which happen.
+    pub fn try_finish(&mut self) -> io::Result<()> {
+        try!(self.inner.finish());
+        Ok(())
+    }
+
+    /// Consumes this decoder, flushing the output stream.
+    ///
+    /// This will flush the underlying data stream and then return the contained
+    /// writer if the flush succeeded.
+    ///
+    /// Note that this function may not be suitable to call in a situation where
+    /// the underlying stream is an asynchronous I/O stream. To finish a stream
+    /// the `try_finish` (or `shutdown`) method should be used instead. To
+    /// re-acquire ownership of a stream it is safe to call this method after
+    /// `try_finish` or `shutdown` has returned `Ok`.
+    ///
+    /// # Errors
+    ///
+    /// This function will perform I/O to complete this stream, and any I/O
+    /// errors which occur will be returned from this function.
+    pub fn finish(mut self) -> io::Result<W> {
+        try!(self.inner.finish());
+
+        if self.crc_bytes.len() != 8 {
+            return Err(corrupt())
+        }
+
+        let crc = ((self.crc_bytes[0] as u32) << 0)
+            | ((self.crc_bytes[1] as u32) << 8)
+            | ((self.crc_bytes[2] as u32) << 16)
+            | ((self.crc_bytes[3] as u32) << 24);
+        let amt = ((self.crc_bytes[4] as u32) << 0)
+            | ((self.crc_bytes[5] as u32) << 8)
+            | ((self.crc_bytes[6] as u32) << 16)
+            | ((self.crc_bytes[7] as u32) << 24);
+        if crc != self.inner.get_ref().crc().sum() as u32 {
+            return Err(corrupt());
+        }
+        if amt != self.inner.get_ref().crc().amount() {
+            return Err(corrupt());
+        }
+
+        Ok(self.inner.take_inner().into_inner())
+    }
+
+    fn write_buf(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let (n, status) = try!(self.inner.write_with_status(buf));
+
+        if status == Status::StreamEnd {
+            if n < buf.len() && self.crc_bytes.len() < 8 {
+                let remaining = buf.len() - n;
+                let crc_bytes = cmp::min(remaining, CRC_BYTES_LEN - self.crc_bytes.len());
+                self.crc_bytes.extend(&buf[n..n+crc_bytes]);
+                return Ok(n+crc_bytes)
+            }
+        }
+        Ok(n)
+    }
+}
+
+impl<W: Write> Write for GzDecoder<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if self.header.is_none() {
+            // trying to avoid buffer usage
+            let (res, pos) = if self.header_buf.get_ref().is_empty() {
+                let mut cur = io::Cursor::new(buf);
+                let res = read_gz_header(&mut cur);
+                (res, cur.position() as usize)
+            } else {
+                self.header_buf.get_mut().extend(buf);
```

Comment:
Hm I think this may hit the same pathology as before (but slightly different). Now if you write one byte (fill the buffer a bit) and then write a megabyte it'll move the entire megabyte into this buffer I think.

Reply:
I do not think buffer solution can solve this problem. but I am not sure I am ready to write state machine for header parsing :)

Reply:
Perhaps the `read_gz_header` could be rewritten to take two slices? That way you could buffer up data if it's not successful and otherwise you'd just pass in the caller's buffer directly

Reply:
ok, I'll check what I can do